### DISCUSSION TITLE
collection/arraydeque

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#195] [Announcements]
# collection/arraydeque

本文详细解析了 Java 中的双端队列 ArrayDeque 的实现原理、功能特点以及源码，为您提供了 ArrayDeque 的实际应用示例和性能优化建议。阅读本文，将帮助您更深入地理解双端队列在 Java 编程中的应用，从而在实际编程中充分发挥其优势。

https://tobebetterjavaer.com/collection/arraydeque.html

**Comment by independenter** [comment]
最新的java17有变动，扩容代码变成了小于64是，此次扩容两个，否则扩容1.5倍
private void grow(int needed) {
        // overflow-conscious code
        final int oldCapacity = elements.length;
        int newCapacity;
        // Double capacity if small; else grow by 50%
        int jump = (oldCapacity < 64) ? (oldCapacity + 2) : (oldCapacity >> 1);
        if (jump < needed
            || (newCapacity = (oldCapacity + jump)) - MAX_ARRAY_SIZE > 0)
            newCapacity = newCapacity(needed, jump);
        final Object[] es = elements = Arrays.copyOf(elements, newCapacity);
        // Exceptionally, here tail == head needs to be disambiguated
        if (tail < head || (tail == head && es[head] != null)) {
            // wrap around; slide first leg forward to end of array
            int newSpace = newCapacity - oldCapacity;
            System.arraycopy(es, head,
                             es, head + newSpace,
                             oldCapacity - head);
            for (int i = head, to = (head += newSpace); i < to; i++)
                es[i] = null;
        }
    }

**Comment by white0820** [comment]
在方法剖析第一个addFirst的最后，tail=n，并不是像文中描述的n是新容量的大小，而是原来数组容量的大小，因为tail是一直要指向已用数组的尾部的。

**Comment by Eivae** [comment]
讲addFirst下标越界那里，为什么element.length一定要为2的指数倍？

**Comment by ustcjsj** [comment]
对应表是错的，不需要看了

**Comment by ayu255** [comment]
// 遍历队列查找元素
Iterator<String> iterator = stack.iterator();
int index = -1;
String target = "王二";
while (iterator.hasNext()) {
    String element = iterator.next();
    index++;
    if (element.equals(target)) {
        break;
    }
}

if (index == -1) {
    System.out.println("元素 " + target + " 不存在于队列中");
} else {
    System.out.println("元素 " + target + " 在队列中的位置为：" + index);
}
这个错了吧，
